### PR TITLE
Introduce deref cmd

### PR DESCRIPTION
**Pull request by amir73il**
**deref** - copy changes from `upperdir` to `uppernew` while unfolding redirect directories and metacopy regular files, so that new upperdir is compatible with legacy overlayfs driver.

### CODE REVIEW

**Comment by kmxz** [comment]
Thanks for the PR! I'll take a look quickly.

File: logic.c [hunk 1]
```diff
@@ -202,140 +333,163 @@ int list_deleted_files(const char *path, size_t root_path_len, bool verbose) { /
...
 }
 
-int merge_d(const char *lower_path, const char* upper_path, const size_t lower_root_len, const struct stat *lower_status, const struct stat *upper_status, bool verbose, FILE* script_stream, int *fts_instr) {
+static int merge_d(const char *lower_path, const char* upper_path, const size_t lower_root_len, const struct stat *lower_status, const struct stat *upper_status, FILE* script_stream, int *fts_instr) {
+    bool redirect;
+    if (is_redirect(upper_path, &redirect) < 0) { return -1; }
+    // merging redirects is not supported, we must abort merge so redirected lower (under whiteout) won't be deleted
+    // upper_path may be hiding the directory in lower_path, but there may be another redirect upper pointing at it
+    if (redirect) {
+        fprintf(stderr, "Found redirect on %s. merging redirect is not supported - Abort.\n", upper_path);
```
Comment (kmxz) [inline logic.c]: nit: merging -> Merging?

File: main.c [hunk 2]
```diff
@@ -175,15 +216,24 @@ int main(int argc, char *argv[]) {
...
+            out = merge(lower, upper, script);
+        } else if (strcmp(argv[optind], "deref") == 0) {
+            if (!mnt || !vars[UPPERNEW]) { fprintf(stderr, "'deref' command requires --uppernew and --mountdir.\n"); return EXIT_FAILURE; }
+            if (!directory_exists(mnt)) {
+                fprintf(stderr, "OverlayFS mount directory cannot be opened.\n");
+                goto see_help;
+            }
+            script = create_shell_script(filename_template);
+            if (script == NULL) { fprintf(stderr, "Script file cannot be created.\n"); return EXIT_FAILURE; }
+	    out = deref(mnt, upper, script);
```
Comment (kmxz) [inline main.c]: nit: two more spaces?

File: sh.c [hunk 3]
```diff
@@ -17,10 +27,28 @@ FILE* create_shell_script(char *tmp_path_buffer) {
     time_t rawtime;
     time (&rawtime);
     fprintf(f, "# This shell script is generated by overlayfs-tools on %s\n", ctime (&rawtime));
+    for (int i=0; i < NUM_VARS; i++)
+        if (vars[i])
+            fprintf(f, "%s='%s'\n", var_names[i], vars[i]);
```
Comment (kmxz) [inline sh.c]: It may not work well when `vars[i]` contains single quotes.

**Review by kmxz — CHANGES_REQUESTED** [review]
Thanks for the massive improvements!

I'm not very familiar with the updates in OverlayFS, so I'm not confident in proofreading the logic. But the code surely looks great. I'll definitely merge it.

File: sh.c (hunk 3 above)
Comment (amir73il) [reply sh.c]: in main.c vars[] is assigned a string returned from realpath(3). It does not contain any quotes, even if quotes are passed in command line

Comment (kmxz) [reply sh.c]: I mean the case where the actual directory name itself include quotes.

Comment (amir73il) [reply sh.c]: Can you show this with an example?
Actually run overlay that generates a bad script?

Comment (kmxz) [reply sh.c]: Here's an example: I set up it with `mkdir "/tmp/test's upper"; mkdir "/tmp/test's lower"; touch "/tmp/test's upper/foo"`, and run `sudo ./overlay merge -l /tmp/test\'s\ lower/ -u /tmp/test\'s\ upper/`.

The generated shell script looks like:

```
LOWERDIR='/tmp/test's lower'
UPPERDIR='/tmp/test's upper'
rm -rf $LOWERDIR'/foo.txt'
```

The the single quote in `test's` breaks the string literal, and the script fails to execute.

Comment (amir73il) [reply sh.c]: I see. I am not an expect in quoting as you can see.
So I need to fix create_shell_script() to use quote().
Let me try to fix that.

Comment (amir73il) [reply sh.c]: done. please verify I did not miss anything

**Comment by kmxz** [comment]
Merged. Much appreciation for the great contribution!

**Pull request merged**